File: typescript/src/http.ts
import { TachyonConnectionError, TachyonTimeoutError, errorFromResponse } from './errors.js';

export interface HttpClientOptions {
  /** Base URL of the Tachyon server, e.g. `http://localhost:8108`. */
  url: string;
  /** Sent as `X-TACHYON-API-KEY` on every request. */
  apiKey?: string;
  /** Per-request timeout in milliseconds. Default 15000. */
  timeoutMs?: number;
  /** Extra headers merged into every request. */
  headers?: Record<string, string>;
  /** Override the `fetch` implementation (mainly for testing). */
  fetch?: typeof fetch;
}

interface RequestOptions {
  query?: Record<string, unknown>;
  body?: unknown;
}

interface RawResponse {
  status: number;
  text: string;
}

/** Thin JSON-over-HTTP client shared by every resource in the SDK. */
export class HttpClient {
  private readonly baseUrl: string;
  private readonly apiKey: string | undefined;
  private readonly timeoutMs: number;
  private readonly extraHeaders: Record<string, string>;
  private readonly fetchImpl: typeof fetch;

  constructor(options: HttpClientOptions) {
    if (!options.url) {
      throw new Error('Tachyon client requires a `url`.');
    }
    this.baseUrl = options.url.replace(/\/+$/, '');
    this.apiKey = options.apiKey;
    this.timeoutMs = options.timeoutMs ?? 15_000;
    this.extraHeaders = options.headers ?? {};

    const fetchImpl = options.fetch ?? globalThis.fetch;
    if (!fetchImpl) {
      throw new Error(
        'No `fetch` implementation is available in this runtime. Upgrade to Node 18+ ' +
          'or pass one explicitly via the `fetch` client option.',
      );
    }
    this.fetchImpl = fetchImpl;
  }

  async request<T>(method: string, path: string, opts: RequestOptions = {}): Promise<T> {
    const { status, text } = await this.send(method, path, opts);
    if (status === 204 || text.length === 0) {
      return undefined as T;
    }

    let payload: unknown;
    try {
      payload = JSON.parse(text);
    } catch {
      if (status >= 400) {
        throw errorFromResponse(status, text, 'internal_error');
      }
      throw new TachyonConnectionError(`Tachyon returned a non-JSON response: ${truncate(text)}`);
    }

    if (status >= 400) {
      throw errorFromResponse(status, extractMessage(payload, text), extractCode(payload));
    }

    return payload as T;
  }

  async requestText(method: string, path: string, opts: RequestOptions = {}): Promise<string> {
    const { status, text } = await this.send(method, path, opts);
    if (status >= 400) {
      let payload: unknown;
      try {
        payload = JSON.parse(text);
      } catch {
        throw errorFromResponse(status, text || `HTTP ${status}`, 'internal_error');
      }
      throw errorFromResponse(status, extractMessage(payload, text), extractCode(payload));
    }
    return text;
  }

  private async send(method: string, path: string, opts: RequestOptions): Promise<RawResponse> {
    const url = this.buildUrl(path, opts.query);
    const headers: Record<string, string> = { Accept: 'application/json', ...this.extraHeaders };
    if (this.apiKey) {
      headers['X-TACHYON-API-KEY'] = this.apiKey;
    }

    let body: string | undefined;
    if (opts.body !== undefined) {
      headers['Content-Type'] = 'application/json';
      body = JSON.stringify(opts.body);
    }

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeoutMs);
    try {
      const response = await this.fetchImpl(url, { method, headers, body, signal: controller.signal });
      const text = await response.text();
      return { status: response.status, text };
    } catch (err) {
      if (err instanceof Error && err.name === 'AbortError') {
        throw new TachyonTimeoutError(`Request to ${url} timed out after ${this.timeoutMs}ms`);
      }
      throw new TachyonConnectionError(
        `Failed to reach Tachyon at ${url}: ${err instanceof Error ? err.message : String(err)}`,
        err,
      );
    } finally {
      clearTimeout(timer);
    }
  }

  private buildUrl(path: string, query?: Record<string, unknown>): string {
    const url = new URL(this.baseUrl + path);
    if (query) {
      for (const [key, value] of Object.entries(query)) {
        if (value === undefined || value === null) continue;
        url.searchParams.set(key, String(value));
      }
    }
    return url.toString();
  }
}

function extractMessage(payload: unknown, fallback: string): string {
  if (payload && typeof payload === 'object' && 'error' in payload) {
    const error = (payload as { error?: unknown }).error;
    if (error && typeof error === 'object' && 'message' in error && typeof (error as { message?: unknown }).message === 'string') {
      return (error as { message: string }).message;
    }
  }
  return fallback;
}

function extractCode(payload: unknown): string {
  if (payload && typeof payload === 'object' && 'error' in payload) {
    const error = (payload as { error?: unknown }).error;
    if (error && typeof error === 'object' && 'code' in error && typeof (error as { code?: unknown }).code === 'string') {
      return (error as { code: string }).code;
    }
  }
  return 'internal_error';
}

function truncate(text: string, max = 200): string {
  return text.length > max ? `${text.slice(0, max)}…` : text;
}

File: go/http.go
package tachyon

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
)

const apiKeyHeader = "X-TACHYON-API-KEY"

// httpTransport is the thin JSON-over-HTTP client shared by every service in the SDK.
type httpTransport struct {
	baseURL    string
	apiKey     string
	headers    map[string]string
	httpClient *http.Client
}

type requestOptions struct {
	query map[string]string
	body  any
}

type apiErrorBody struct {
	Error struct {
		Code    string `json:"code"`
		Message string `json:"message"`
	} `json:"error"`
}

// do sends a request and decodes a JSON response into out (which may be nil
// for responses with no body, e.g. DELETE).
func (t *httpTransport) do(ctx context.Context, method, path string, opts requestOptions, out any) error {
	status, body, err := t.send(ctx, method, path, opts)
	if err != nil {
		return err
	}
	if status == 204 || len(body) == 0 {
		return nil
	}

	if status >= 400 {
		return errorFromBody(status, body)
	}

	if out == nil {
		return nil
	}
	if err := json.Unmarshal(body, out); err != nil {
		return &ConnectionError{
			Message: fmt.Sprintf("tachyon returned a response that could not be decoded: %v", err),
			Err:     err,
		}
	}
	return nil
}

// doText sends a request and returns the raw response body as text (used for /metrics).
func (t *httpTransport) doText(ctx context.Context, method, path string) (string, error) {
	status, body, err := t.send(ctx, method, path, requestOptions{})
	if err != nil {
		return "", err
	}
	if status >= 400 {
		return "", errorFromBody(status, body)
	}
	return string(body), nil
}

func errorFromBody(status int, body []byte) error {
	var payload apiErrorBody
	if err := json.Unmarshal(body, &payload); err != nil || payload.Error.Code == "" {
		return errorFromResponse(status, string(body), "internal_error")
	}
	return errorFromResponse(status, payload.Error.Message, payload.Error.Code)
}

func (t *httpTransport) send(ctx context.Context, method, path string, opts requestOptions) (int, []byte, error) {
	u, err := url.Parse(t.baseURL + path)
	if err != nil {
		return 0, nil, fmt.Errorf("invalid URL %q: %w", t.baseURL+path, err)
	}
	if len(opts.query) > 0 {
		q := u.Query()
		for k, v := range opts.query {
			q.Set(k, v)
		}
		u.RawQuery = q.Encode()
	}

	var bodyReader io.Reader
	if opts.body != nil {
		encoded, err := json.Marshal(opts.body)
		if err != nil {
			return 0, nil, fmt.Errorf("failed to encode request body: %w", err)
		}
		bodyReader = bytes.NewReader(encoded)
	}

	req, err := http.NewRequestWithContext(ctx, method, u.String(), bodyReader)
	if err != nil {
		return 0, nil, fmt.Errorf("failed to build request: %w", err)
	}
	req.Header.Set("Accept", "application/json")
	if opts.body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	if t.apiKey != "" {
		req.Header.Set(apiKeyHeader, t.apiKey)
	}
	for k, v := range t.headers {
		req.Header.Set(k, v)
	}

	resp, err := t.httpClient.Do(req)
	if err != nil {
		if errors.Is(err, context.DeadlineExceeded) {
			return 0, nil, &ConnectionError{
				Message: fmt.Sprintf("request to %s timed out: %v", u.String(), err),
				Err:     err,
				Timeout: true,
			}
		}
		return 0, nil, &ConnectionError{
			Message: fmt.Sprintf("failed to reach tachyon at %s: %v", u.String(), err),
			Err:     err,
		}
	}
	defer resp.Body.Close()

	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return 0, nil, &ConnectionError{
			Message: fmt.Sprintf("failed to read response from %s: %v", u.String(), err),
			Err:     err,
		}
	}

	return resp.StatusCode, data, nil
}
